In [1]:
import pandas as pd
import numpy as np
import scipy
import sklearn

In [2]:
excel_origin=pd.ExcelFile('MCM_NFLIS_Data.xlsx')# 原始数据
data_origin=pd.read_excel(excel_origin,'Data')

In [3]:
data_origin.drop('State',axis=1,inplace=True)
data_origin.drop('COUNTY',axis=1,inplace=True)

In [4]:
data_origin

YYYY  FIPS_State  FIPS_County  FIPS_Combined  \
0      2010          21            1          21001   
1      2010          21            1          21001   
2      2010          21            3          21003   
3      2010          21            3          21003   
4      2010          21            3          21003   
5      2010          21            3          21003   
6      2010          21            5          21005   
7      2010          21            5          21005   
8      2010          21            5          21005   
9      2010          21            5          21005   
10     2010          21            5          21005   
11     2010          21            7          21007   
12     2010          21            9          21009   
13     2010          21            9          21009   
14     2010          21            9          21009   
15     2010          21            9          21009   
16     2010          21            9          21009   
17     2010          21            9          21009   
18     2010          21            9          21009   
19     2010          21            9          21009   
20     2010          21            9          21009   
21     2010          21           11          21011   
22     2010          21           11          21011   
23     2010          21           11          21011   
24     2010          21           11          21011   
25     2010          21           11          21011   
26     2010          21           13          21013   
27     2010          21           13          21013   
28     2010          21           13          21013   
29     2010          21           13          21013   
...     ...         ...          ...            ...   
24032  2017          54           83          54083   
24033  2017          54           83          54083   
24034  2017          54           83          54083   
24035  2017          54           87          54087   
24036  2017          54           93          54093   
24037  2017          54           93          54093   
24038  2017          54           97          54097   
24039  2017          54           97          54097   
24040  2017          54           99          54099   
24041  2017          54           99          54099   
24042  2017          54           99          54099   
24043  2017          54           99          54099   
24044  2017          54           99          54099   
24045  2017          54          101          54101   
24046  2017          54          101          54101   
24047  2017          54          101          54101   
24048  2017          54          103          54103   
24049  2017          54          107          54107   
24050  2017          54          107          54107   
24051  2017          54          107          54107   
24052  2017          54          107          54107   
24053  2017          54          107          54107   
24054  2017          54          107          54107   
24055  2017          54          107          54107   
24056  2017          54          107          54107   
24057  2017          54          107          54107   
24058  2017          54          107          54107   
24059  2017          54          107          54107   
24060  2017          54          107          54107   
24061  2017          54          109          54109   

                    SubstanceName  DrugReports  TotalDrugReportsCounty  \
0                     Hydrocodone            6                      88   
1                       Methadone            1                      88   
2                     Hydrocodone           10                     168   
3                       Methadone            4                     168   
4                       Oxycodone           15                     168   
5                     Oxymorphone            1                     168   
6                   Buprenorphine            1                      98   
7   

In [5]:
names=locals()
for i in range(18):
    names['data_201'+str(i-1)]=data_origin.iloc[:,:][data_origin.YYYY<2010+i]
    names['data_201'+str(i-1)]=names['data_201'+str(i-1)].iloc[:,:][data_origin.YYYY>2008+i]
# 提取出了各年份的数据

E:\anaconda\lib\site-packages\ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


In [84]:
len(data_2010)

2836

In [78]:
for i in range(8):
    names['dataxinde_201'+str(i)]=data_origin.iloc[:,:][data_origin.YYYY<2010+i]
    names['dataxinde_201'+str(i)]=names['data_201'+str(i-1)].iloc[:,:][data_origin.YYYY>2008+i]
# 提取出了各年份的数据

E:\anaconda\lib\site-packages\ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
dataxinde_201

In [80]:
names['dataxinde_201'+str(i)]['SubstanceName']

16880    Buprenorphine
16881      Hydrocodone
16882    Hydromorphone
16883         Morphine
16884    Buprenorphine
16885           Heroin
16886      Hydrocodone
16887         Morphine
16888        Oxycodone
16889         Fentanyl
16890           Heroin
16891      Hydrocodone
16892         Morphine
16893        Oxycodone
16894        Methadone
16895    Buprenorphine
16896           Heroin
16897      Hydrocodone
16898    Hydromorphone
16899         Morphine
16900        Oxycodone
16901    Buprenorphine
16902         Fentanyl
16903           Heroin
16904      Hydrocodone
16905         Morphine
16906        Oxycodone
16907    Buprenorphine
16908          Codeine
16909           Heroin
             ...      
20179      Hydrocodone
20180        Oxycodone
20181    Buprenorphine
20182    Buprenorphine
20183      Hydrocodone
20184    Hydromorphone
20185        Oxycodone
20186    Buprenorphine
20187      Hydrocodone
20188        Oxycodone
20189    Buprenorphine
20190         Fentanyl
20191      

In [6]:
data_2010

##### 2010年郡数

county={}
for i in range(2836):
    FIPS_Combined=data_2010.iloc[i-1][3]
    #print(FIPS_Combined)
    if FIPS_Combined not in county.keys():
        county[FIPS_Combined]=FIPS_Combined
len(county)

445

In [7]:
data_origin

YYYY  FIPS_State  FIPS_County  FIPS_Combined  \
0      2010          21            1          21001   
1      2010          21            1          21001   
2      2010          21            3          21003   
3      2010          21            3          21003   
4      2010          21            3          21003   
5      2010          21            3          21003   
6      2010          21            5          21005   
7      2010          21            5          21005   
8      2010          21            5          21005   
9      2010          21            5          21005   
10     2010          21            5          21005   
11     2010          21            7          21007   
12     2010          21            9          21009   
13     2010          21            9          21009   
14     2010          21            9          21009   
15     2010          21            9          21009   
16     2010          21            9          21009   
17     2010          21            9          21009   
18     2010          21            9          21009   
19     2010          21            9          21009   
20     2010          21            9          21009   
21     2010          21           11          21011   
22     2010          21           11          21011   
23     2010          21           11          21011   
24     2010          21           11          21011   
25     2010          21           11          21011   
26     2010          21           13          21013   
27     2010          21           13          21013   
28     2010          21           13          21013   
29     2010          21           13          21013   
...     ...         ...          ...            ...   
24032  2017          54           83          54083   
24033  2017          54           83          54083   
24034  2017          54           83          54083   
24035  2017          54           87          54087   
24036  2017          54           93          54093   
24037  2017          54           93          54093   
24038  2017          54           97          54097   
24039  2017          54           97          54097   
24040  2017          54           99          54099   
24041  2017          54           99          54099   
24042  2017          54           99          54099   
24043  2017          54           99          54099   
24044  2017          54           99          54099   
24045  2017          54          101          54101   
24046  2017          54          101          54101   
24047  2017          54          101          54101   
24048  2017          54          103          54103   
24049  2017          54          107          54107   
24050  2017          54          107          54107   
24051  2017          54          107          54107   
24052  2017          54          107          54107   
24053  2017          54          107          54107   
24054  2017          54          107          54107   
24055  2017          54          107          54107   
24056  2017          54          107          54107   
24057  2017          54          107          54107   
24058  2017          54          107          54107   
24059  2017          54          107          54107   
24060  2017          54          107          54107   
24061  2017          54          109          54109   

                    SubstanceName  DrugReports  TotalDrugReportsCounty  \
0                     Hydrocodone            6                      88   
1                       Methadone            1                      88   
2                     Hydrocodone           10                     168   
3                       Methadone            4                     168   
4                       Oxycodone           15                     168   
5                     Oxymorphone            1                     168   
6                   Buprenorphine            1                      98   
7   

In [8]:
county2={}
for ii in range(24062):# ii 是从0开始的
    FIPS_Combined=data_origin.iloc[ii][3]
    if FIPS_Combined not in county2.keys():
        county2[FIPS_Combined]=FIPS_Combined

In [9]:
# 阿片类药物数
opioids={}
for ii in range(24062):# ii 是从0开始的
    SubstanceName=data_origin.iloc[ii][4]
    if SubstanceName not in opioids.keys():
        opioids[SubstanceName]=SubstanceName

In [81]:
opioids

{'Hydrocodone': 'Hydrocodone',
 'Methadone': 'Methadone',
 'Oxycodone': 'Oxycodone',
 'Oxymorphone': 'Oxymorphone',
 'Buprenorphine': 'Buprenorphine',
 'Morphine': 'Morphine',
 'Hydromorphone': 'Hydromorphone',
 'Meperidine': 'Meperidine',
 'Tramadol': 'Tramadol',
 'Codeine': 'Codeine',
 'Fentanyl': 'Fentanyl',
 'Heroin': 'Heroin',
 'Propoxyphene': 'Propoxyphene',
 'Methorphan': 'Methorphan',
 'Pentazocine': 'Pentazocine',
 'Pethidine': 'Pethidine',
 'Thebaine': 'Thebaine',
 'Dextropropoxyphene': 'Dextropropoxyphene',
 'Opiates': 'Opiates',
 'Dihydrocodeine': 'Dihydrocodeine',
 'Dihydromorphone': 'Dihydromorphone',
 'Acetylcodeine': 'Acetylcodeine',
 'Alphaprodine': 'Alphaprodine',
 'Opium': 'Opium',
 'Butorphanol': 'Butorphanol',
 'Desmethylprodine ': 'Desmethylprodine ',
 'Mitragynine': 'Mitragynine',
 'ANPP': 'ANPP',
 'Nalbuphine': 'Nalbuphine',
 'Remifentanil': 'Remifentanil',
 'Acetyldihydrocodeine': 'Acetyldihydrocodeine',
 'Hydrocodeinone': 'Hydrocodeinone',
 'MT-45': 'MT-45',
 

In [10]:
len(opioids)# 共69中合成物

69

In [11]:
len(county2)

461

似乎有的年份有的郡的数据是缺失的，总共的郡的出现的个数为461,2010年只有445个郡出现。论文中提到总郡数为462，不知道为什么与我的实验数据不对。论文考虑到总共462个郡的数据过大，通过地理位置（经纬度）采用knn算法聚类为100个区域，不知道这样是否科学，在这么短的时间找到经纬度数据集可谓说是很厉害了，而且这还是个前期工作。

引知乎回答：[如何评价2019美赛C题？ - 知乎](https://www.zhihu.com/question/310225524/answer/583190402)，似乎有数据，但仔细查却发现是city的数据，不能用作郡的数据，提供的有zip的，但是要经过繁琐的处理

In [12]:
UScities_excel_origin=pd.ExcelFile('uszips.xlsx')# 原始数据
uscities_origin=pd.read_excel(UScities_excel_origin,'Sheet1')

得到新的frame后会发现index是不连续的，这是因为pandas在设计的时候考虑到安全问题保持index不可修改，方便数据在多个frame之间传递，如果我们想要更改index，可以用`.reset_index(drop=True)`方法，如下：

In [85]:
#USAcityLonLat=USAcityLonLat.reset_index(drop=True)
import re

county_lonAndLat={}
for ii in range(24062):# ii 是从0开始的
    FIPS_Combined=data_origin.iloc[ii][3]
    if FIPS_Combined not in county_lonAndLat.keys():
        USAcityLonLat=uscities_origin.loc[uscities_origin.county_fips==FIPS_Combined]
        if USAcityLonLat.shape[0]!=0:
            USAcityLonLat=USAcityLonLat.reset_index(drop=True)
            lat=USAcityLonLat.iloc[0,1]
            lng=USAcityLonLat.iloc[0,2]
            county_lonAndLat[FIPS_Combined]=(lat,lng)
        else:# 失效时换用字符串匹配
            FIPS_Combined_pattern=str(FIPS_Combined)
            for j in range(33099):
                uscities_fisp_str=str(uscities_origin.iloc[j,14])
                if re.search(FIPS_Combined_pattern,uscities_fisp_str):
                    lat=uscities_origin.iloc[j,1]
                    lng=uscities_origin.iloc[j,2]
                    county_lonAndLat[FIPS_Combined]=(lat,lng)
                    break


解释一下上面这部分代码，由于MCM_NFLIS_Data.xlsx中给的数据是county数据而非城市数据，不少fips编号都是县级的，用uscities.xlsx无法直接对应，改用uszip.xlsx，而且很多county的fips码是被它的上级城市所覆盖的，就如由于MCM_NFLIS_Data的BEDFORD CITY，fips:51515，在city数据集中找不到对应关系，似乎因为2015年bedford city 和bedford county合并，共同采用51019。但是fips_all一列中仍会包含原来采用的fips码，形如"51019|51515"，这时候需要再加上一个模式匹配算法。

最后得到一个包含了地区fips码及经纬度的dataframe，用其进行聚类：

In [91]:
data_origin

YYYY  FIPS_State  FIPS_County  FIPS_Combined  \
0      2010          21            1          21001   
1      2010          21            1          21001   
2      2010          21            3          21003   
3      2010          21            3          21003   
4      2010          21            3          21003   
5      2010          21            3          21003   
6      2010          21            5          21005   
7      2010          21            5          21005   
8      2010          21            5          21005   
9      2010          21            5          21005   
10     2010          21            5          21005   
11     2010          21            7          21007   
12     2010          21            9          21009   
13     2010          21            9          21009   
14     2010          21            9          21009   
15     2010          21            9          21009   
16     2010          21            9          21009   
17     2010          21            9          21009   
18     2010          21            9          21009   
19     2010          21            9          21009   
20     2010          21            9          21009   
21     2010          21           11          21011   
22     2010          21           11          21011   
23     2010          21           11          21011   
24     2010          21           11          21011   
25     2010          21           11          21011   
26     2010          21           13          21013   
27     2010          21           13          21013   
28     2010          21           13          21013   
29     2010          21           13          21013   
...     ...         ...          ...            ...   
24032  2017          54           83          54083   
24033  2017          54           83          54083   
24034  2017          54           83          54083   
24035  2017          54           87          54087   
24036  2017          54           93          54093   
24037  2017          54           93          54093   
24038  2017          54           97          54097   
24039  2017          54           97          54097   
24040  2017          54           99          54099   
24041  2017          54           99          54099   
24042  2017          54           99          54099   
24043  2017          54           99          54099   
24044  2017          54           99          54099   
24045  2017          54          101          54101   
24046  2017          54          101          54101   
24047  2017          54          101          54101   
24048  2017          54          103          54103   
24049  2017          54          107          54107   
24050  2017          54          107          54107   
24051  2017          54          107          54107   
24052  2017          54          107          54107   
24053  2017          54          107          54107   
24054  2017          54          107          54107   
24055  2017          54          107          54107   
24056  2017          54          107          54107   
24057  2017          54          107          54107   
24058  2017          54          107          54107   
24059  2017          54          107          54107   
24060  2017          54          107          54107   
24061  2017          54          109          54109   

                    SubstanceName  DrugReports  TotalDrugReportsCounty  \
0                     Hydrocodone            6                      88   
1                       Methadone            1                      88   
2                     Hydrocodone           10                     168   
3                       Methadone            4                     168   
4                       Oxycodone           15                     168   
5                     Oxymorphone            1                     168   
6                   Buprenorphine            1                      98   
7   

In [15]:
model=sc.KMeans(n_clusters=100)
model.fit(county_lonAndLatFram)
# pandas数据可以直接放入？不太清楚，能聚就行，以后再分析

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=100, n_init=10, n_jobs=1, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0)

In [16]:
model.labels_

array([10, 55, 68,  3, 55, 92, 18, 77, 35, 93, 29, 35, 27, 49, 68, 12, 95,
       95, 77, 80, 47,  6, 29, 26, 92, 18, 10, 95, 10, 63, 55,  6, 92, 96,
       35, 46, 68, 80,  1, 29,  1,  3, 49, 40,  6, 49, 18, 35, 40, 63,  1,
       26, 61, 47, 96, 46, 77, 46, 61, 40, 61, 46, 27, 18, 66,  6, 29, 95,
       94, 95, 92, 46, 40, 95, 46, 35,  3, 61, 63, 49, 27, 96, 55, 55, 92,
       27, 12, 40, 35, 12, 47,  1, 27, 35, 27, 46, 92, 61, 35, 29, 27, 10,
       96, 68, 94, 68, 40, 94, 26, 47, 12, 68, 10, 63, 61, 27, 96, 35, 45,
       34, 48, 74, 45, 32, 35, 77, 70,  8,  8, 77, 50, 39, 14, 65, 34, 30,
       87, 65, 34, 81, 98, 41, 87, 28,  4, 50, 89, 77, 19,  8, 19, 98, 79,
       14, 34, 79, 32, 14,  4, 93, 81,  8, 34, 19,  8, 70, 65, 34, 28, 30,
       50, 32, 50, 74, 65, 14, 89, 19, 45, 81, 41, 79, 70, 77, 45, 65, 79,
       19,  6, 19, 30, 70,  4,  4, 89,  8, 45, 79, 50, 74, 34, 87, 19, 19,
       71, 13, 90, 39, 37, 11, 37,  5, 11, 13, 69, 23, 33, 37, 11, 13, 69,
       23, 85, 48, 71, 21

In [20]:
county_lonAndLatFram.loc[21001,'zone']

10

In [18]:
county_lonAndLatFram['zone']=model.labels_
county_lonAndLatFram

0         1  zone
21001  37.18414 -85.01405    10
21003  36.67111 -86.26973    55
21005  38.00322 -84.98494    68
21007  37.14590 -88.94554     3
21009  36.82162 -85.98925    55
21011  38.05909 -83.67994    92
21013  36.72772 -83.59301    18
21015  39.00803 -84.76746    77
21017  38.28738 -84.13209    35
21019  38.47219 -82.64599    93
21021  37.63785 -84.80074    29
21023  38.73062 -83.98266    35
21025  37.44763 -83.18914    27
21027  37.77311 -86.62818    49
21029  38.03973 -85.55479    68
21031  37.21978 -86.66900    12
21033  37.20307 -88.01808    95
21035  36.69232 -88.29298    95
21037  38.91491 -84.40265    77
21039  36.80481 -89.02655    80
21041  38.66355 -85.16782    47
21043  38.26330 -82.82466     6
21045  37.22715 -84.77861    29
21047  36.99700 -87.66706    26
21049  37.96943 -84.14517    92
21051  36.94428 -83.54395    18
21053  36.73159 -85.13322    10
21055  37.17273 -88.18779    95
21057  36.78188 -85.37894    10
21059  37.74038 -87.24615    63
...         ...       ...   ...
54079  38.50861 -81.83935    25
54081  37.82536 -81.42867    73
54083  38.74033 -80.18293     2
54085  39.06903 -81.18634    17
54087  38.58088 -81.26921    25
54089  37.56075 -80.81113     9
54093  39.07893 -79.41767    22
54095  39.45283 -81.04396    17
54097  39.00049 -80.20387     2
54099  38.39676 -82.55398    93
54103  39.63217 -80.76542    58
54105  39.02869 -81.20434    17
54107  39.23905 -81.57619    74
54109  37.48718 -81.37710    73
21105  36.68755 -88.99122    80
21225  37.66397 -87.93653    63
39067  40.25609 -81.02123    32
51131  37.43369 -75.87307    24
54021  38.84772 -80.94138    17
54027  39.48344 -78.56789    59
54041  39.05153 -80.68267    58
54071  38.62752 -79.19655    60
54073  39.36798 -81.28758    17
54091  39.25657 -80.12616    82
54101  38.54090 -80.58270    83
51036  37.34630 -77.05926    76
51097  37.77866 -76.95957    44
21091  37.83487 -86.76309    49
54043  38.25115 -81.78618    25
51580  37.76844 -80.07654    67

[461 rows x 3 columns]

由于调用sklearn库，使得聚类的过程十分简单，按照论文，现在我们需要得到两个frame，一个为根据zone(聚得的100个类)划分，每一个zone的毒品信息，一个为每个zone包含的county和聚类中心的fips编码，原文是写入excel的，dataframe不支持list写入某一个元素，所以就略过第二个框架，不过后期总结要写上datafrme每一个元素支持的类型和xlwt的用法（python写入excel）

In [39]:
zone=pd.Series(np.arange(100),index=np.arange(0,6900,69))
zone=zone.reindex(range(6900), method='ffill')

data_2010_zone=pd.DataFrame(columns=['YYYY','Zone','substanceName','DrugReport','TotalDrugReportZone'])
data_2010_zone['YYYY']=2010*np.ones(6900)
data_2010_zone['Zone']=zone
opioids_series=pd.Series(opioids).reset_index(drop=True)
opioids_series
for i in range(100):
    for j in range(69):
        data_2010_zone.iloc[69*i+j,2]=opioids_series[j]
data_2010_zone['DrugReport']=np.zeros(6900)
data_2010_zone['TotalDrugReportZone']=np.zeros(6900)

In [40]:
'''
else:# 失效时换用字符串匹配
            FIPS_Combined_pattern=str(FIPS_Combined)
            for j in range(33099):
                uscities_fisp_str=str(uscities_origin.iloc[j,14])
                if re.search(FIPS_Combined_pattern,uscities_fisp_str):
                    lat=uscities_origin.iloc[j,1]
                    lng=uscities_origin.iloc[j,2]
                    county_lonAndLat[FIPS_Combined]=(lat,lng)
                    break
'''
data_2010_zone

YYYY  Zone                  substanceName  DrugReport  \
0     2010.0     0                    Hydrocodone         0.0   
1     2010.0     0                      Methadone         0.0   
2     2010.0     0                      Oxycodone         0.0   
3     2010.0     0                    Oxymorphone         0.0   
4     2010.0     0                  Buprenorphine         0.0   
5     2010.0     0                       Morphine         0.0   
6     2010.0     0                  Hydromorphone         0.0   
7     2010.0     0                     Meperidine         0.0   
8     2010.0     0                       Tramadol         0.0   
9     2010.0     0                        Codeine         0.0   
10    2010.0     0                       Fentanyl         0.0   
11    2010.0     0                         Heroin         0.0   
12    2010.0     0                   Propoxyphene         0.0   
13    2010.0     0                     Methorphan         0.0   
14    2010.0     0                    Pentazocine         0.0   
15    2010.0     0                      Pethidine         0.0   
16    2010.0     0                       Thebaine         0.0   
17    2010.0     0             Dextropropoxyphene         0.0   
18    2010.0     0                        Opiates         0.0   
19    2010.0     0                 Dihydrocodeine         0.0   
20    2010.0     0                Dihydromorphone         0.0   
21    2010.0     0                  Acetylcodeine         0.0   
22    2010.0     0                   Alphaprodine         0.0   
23    2010.0     0                          Opium         0.0   
24    2010.0     0                    Butorphanol         0.0   
25    2010.0     0              Desmethylprodine          0.0   
26    2010.0     0                    Mitragynine         0.0   
27    2010.0     0                           ANPP         0.0   
28    2010.0     0                     Nalbuphine         0.0   
29    2010.0     0                   Remifentanil         0.0   
...      ...   ...                            ...         ...   
6870  2010.0    99                        U-47700         0.0   
6871  2010.0    99                 Fluorofentanyl         0.0   
6872  2010.0    99                 Acryl fentanyl         0.0   
6873  2010.0    99           Cyclopropyl fentanyl         0.0   
6874  2010.0    99               3-Methylfentanyl         0.0   
6875  2010.0    99       Tetrahydrofuran fentanyl         0.0   
6876  2010.0    99               p-Fluorofentanyl         0.0   
6877  2010.0    99    4-Fluoroisobutyryl fentanyl         0.0   
6878  2010.0    99               o-Fluorofentanyl         0.0   
6879  2010.0    99      Fluoroisobutyryl fentanyl         0.0   
6880  2010.0    99               Valeryl fentanyl         0.0   
6881  2010.0    99           cis-3-methylfentanyl         0.0   
6882  2010.0    99      p-methoxybutyryl fentanyl         0.0   
6883  2010.0    99         trans-3-Methylfentanyl         0.0   
6884  2010.0    99               3-Fluorofentanyl         0.0   
6885  2010.0    99        Fluorobutyryl fentanyl          0.0   
6886  2010.0    99     Furanyl/3-Furanyl fentanyl         0.0   
6887  2010.0    99         Methoxyacetyl fentanyl         0.0   
6888  2010.0    99           Cyclopentyl fentanyl         0.0   
6889  2010.0    99                        U-48800         0.0   
6890  2010.0    99                        U-49900         0.0   
6891  2010.0    99                 Benzylfentanyl         0.0   
6892  2010.0    99              Crotonyl fentanyl         0.0   
6893  2010.0    99                Phenyl fentanyl         0.0   
6894  2010.0    99  Cyclopropyl/Crotonyl Fentanyl         0.0   
6895  2010.0    99                        U-51754         0.0   
6896  2010.0    99            Isobutyryl fentanyl         0.0   
6897  2010.0    99               4-Methylfentanyl         0.0   
6898  2010.0    99     3,4-Methylenedioxy U-47700         0.0   
6899  2010.0    99                     Meta

In [41]:
for i in range(2836):
    SubstanceName_pattern=data_2010.iloc[i,4]
    FIPS_Combined=data_2010.iloc[i,3]
    zone=county_lonAndLatFram.loc[FIPS_Combined,'zone']
    for j in range(6900):
        zone_Nmae_pattern=data_2010_zone.iloc[j,2]
        zone_Zone_pattern=data_2010_zone.iloc[j,1]
        if zone_Zone_pattern==zone:
            if re.match(SubstanceName_pattern,zone_Nmae_pattern):
                data_2010_zone.iloc[j,3]=data_2010_zone.iloc[j,3]+data_2010.iloc[i,5]

In [93]:
data_2010

YYYY  FIPS_State  FIPS_County  FIPS_Combined       SubstanceName  \
0     2010          21            1          21001         Hydrocodone   
1     2010          21            1          21001           Methadone   
2     2010          21            3          21003         Hydrocodone   
3     2010          21            3          21003           Methadone   
4     2010          21            3          21003           Oxycodone   
5     2010          21            3          21003         Oxymorphone   
6     2010          21            5          21005       Buprenorphine   
7     2010          21            5          21005         Hydrocodone   
8     2010          21            5          21005           Methadone   
9     2010          21            5          21005            Morphine   
10    2010          21            5          21005           Oxycodone   
11    2010          21            7          21007         Hydrocodone   
12    2010          21            9          21009       Buprenorphine   
13    2010          21            9          21009         Hydrocodone   
14    2010          21            9          21009       Hydromorphone   
15    2010          21            9          21009          Meperidine   
16    2010          21            9          21009           Methadone   
17    2010          21            9          21009            Morphine   
18    2010          21            9          21009           Oxycodone   
19    2010          21            9          21009         Oxymorphone   
20    2010          21            9          21009            Tramadol   
21    2010          21           11          21011       Buprenorphine   
22    2010          21           11          21011         Hydrocodone   
23    2010          21           11          21011       Hydromorphone   
24    2010          21           11          21011           Methadone   
25    2010          21           11          21011           Oxycodone   
26    2010          21           13          21013       Buprenorphine   
27    2010          21           13          21013             Codeine   
28    2010          21           13          21013            Fentanyl   
29    2010          21           13          21013              Heroin   
...    ...         ...          ...            ...                 ...   
2806  2010          54           87          54087           Oxycodone   
2807  2010          54           89          54089              Heroin   
2808  2010          54           89          54089         Hydrocodone   
2809  2010          54           93          54093              Heroin   
2810  2010          54           95          54095              Heroin   
2811  2010          54           95          54095         Hydrocodone   
2812  2010          54           97          54097       Buprenorphine   
2813  2010          54           97          54097            Fentanyl   
2814  2010          54           97          54097              Heroin   
2815  2010          54           97          54097         Hydrocodone   
2816  2010          54           97          54097       Hydromorphone   
2817  2010          54           97          54097            Morphine   
2818  2010          54           97          54097           Oxycodone   
2819  2010          54           99          54099       Buprenorphine   
2820  2010          54           99          54099  Dextropropoxyphene   
2821  2010          54           99          54099         Hydrocodone   
2822  2010          54           99          54099           Oxycodone   
2823  2010          54           99          54099         Oxymorphone   
2824  2010          54          103          54103              Heroin   
2825  2010          54          103          54103         Hydrocodone   
2826  2010          54          103          54103           Oxycodone   
2827  2010          54          105          54105         Hydrocodone   
2828  2010

In [82]:
data_2010

YYYY  FIPS_State  FIPS_County  FIPS_Combined       SubstanceName  \
0     2010          21            1          21001         Hydrocodone   
1     2010          21            1          21001           Methadone   
2     2010          21            3          21003         Hydrocodone   
3     2010          21            3          21003           Methadone   
4     2010          21            3          21003           Oxycodone   
5     2010          21            3          21003         Oxymorphone   
6     2010          21            5          21005       Buprenorphine   
7     2010          21            5          21005         Hydrocodone   
8     2010          21            5          21005           Methadone   
9     2010          21            5          21005            Morphine   
10    2010          21            5          21005           Oxycodone   
11    2010          21            7          21007         Hydrocodone   
12    2010          21            9          21009       Buprenorphine   
13    2010          21            9          21009         Hydrocodone   
14    2010          21            9          21009       Hydromorphone   
15    2010          21            9          21009          Meperidine   
16    2010          21            9          21009           Methadone   
17    2010          21            9          21009            Morphine   
18    2010          21            9          21009           Oxycodone   
19    2010          21            9          21009         Oxymorphone   
20    2010          21            9          21009            Tramadol   
21    2010          21           11          21011       Buprenorphine   
22    2010          21           11          21011         Hydrocodone   
23    2010          21           11          21011       Hydromorphone   
24    2010          21           11          21011           Methadone   
25    2010          21           11          21011           Oxycodone   
26    2010          21           13          21013       Buprenorphine   
27    2010          21           13          21013             Codeine   
28    2010          21           13          21013            Fentanyl   
29    2010          21           13          21013              Heroin   
...    ...         ...          ...            ...                 ...   
2806  2010          54           87          54087           Oxycodone   
2807  2010          54           89          54089              Heroin   
2808  2010          54           89          54089         Hydrocodone   
2809  2010          54           93          54093              Heroin   
2810  2010          54           95          54095              Heroin   
2811  2010          54           95          54095         Hydrocodone   
2812  2010          54           97          54097       Buprenorphine   
2813  2010          54           97          54097            Fentanyl   
2814  2010          54           97          54097              Heroin   
2815  2010          54           97          54097         Hydrocodone   
2816  2010          54           97          54097       Hydromorphone   
2817  2010          54           97          54097            Morphine   
2818  2010          54           97          54097           Oxycodone   
2819  2010          54           99          54099       Buprenorphine   
2820  2010          54           99          54099  Dextropropoxyphene   
2821  2010          54           99          54099         Hydrocodone   
2822  2010          54           99          54099           Oxycodone   
2823  2010          54           99          54099         Oxymorphone   
2824  2010          54          103          54103              Heroin   
2825  2010          54          103          54103         Hydrocodone   
2826  2010          54          103          54103           Oxycodone   
2827  2010          54          105          54105         Hydrocodone   
2828  2010

In [24]:
i=0
SubstanceName_pattern=data_2010.iloc[i,4]
FIPS_Combined=data_2010.iloc[i,3]
zone=county_lonAndLatFram.loc[FIPS_Combined,'zone']

In [42]:
data_2010_zone

YYYY  Zone                  substanceName  DrugReport  \
0     2010.0     0                    Hydrocodone        96.0   
1     2010.0     0                      Methadone        34.0   
2     2010.0     0                      Oxycodone       432.0   
3     2010.0     0                    Oxymorphone         3.0   
4     2010.0     0                  Buprenorphine        47.0   
5     2010.0     0                       Morphine        30.0   
6     2010.0     0                  Hydromorphone        42.0   
7     2010.0     0                     Meperidine         1.0   
8     2010.0     0                       Tramadol        13.0   
9     2010.0     0                        Codeine        17.0   
10    2010.0     0                       Fentanyl         5.0   
11    2010.0     0                         Heroin       304.0   
12    2010.0     0                   Propoxyphene         1.0   
13    2010.0     0                     Methorphan         3.0   
14    2010.0     0                    Pentazocine         0.0   
15    2010.0     0                      Pethidine         0.0   
16    2010.0     0                       Thebaine         0.0   
17    2010.0     0             Dextropropoxyphene         0.0   
18    2010.0     0                        Opiates         0.0   
19    2010.0     0                 Dihydrocodeine         0.0   
20    2010.0     0                Dihydromorphone         0.0   
21    2010.0     0                  Acetylcodeine         0.0   
22    2010.0     0                   Alphaprodine         0.0   
23    2010.0     0                          Opium         1.0   
24    2010.0     0                    Butorphanol         0.0   
25    2010.0     0              Desmethylprodine          0.0   
26    2010.0     0                    Mitragynine         0.0   
27    2010.0     0                           ANPP         0.0   
28    2010.0     0                     Nalbuphine         0.0   
29    2010.0     0                   Remifentanil         0.0   
...      ...   ...                            ...         ...   
6870  2010.0    99                        U-47700         0.0   
6871  2010.0    99                 Fluorofentanyl         0.0   
6872  2010.0    99                 Acryl fentanyl         0.0   
6873  2010.0    99           Cyclopropyl fentanyl         0.0   
6874  2010.0    99               3-Methylfentanyl         0.0   
6875  2010.0    99       Tetrahydrofuran fentanyl         0.0   
6876  2010.0    99               p-Fluorofentanyl         0.0   
6877  2010.0    99    4-Fluoroisobutyryl fentanyl         0.0   
6878  2010.0    99               o-Fluorofentanyl         0.0   
6879  2010.0    99      Fluoroisobutyryl fentanyl         0.0   
6880  2010.0    99               Valeryl fentanyl         0.0   
6881  2010.0    99           cis-3-methylfentanyl         0.0   
6882  2010.0    99      p-methoxybutyryl fentanyl         0.0   
6883  2010.0    99         trans-3-Methylfentanyl         0.0   
6884  2010.0    99               3-Fluorofentanyl         0.0   
6885  2010.0    99        Fluorobutyryl fentanyl          0.0   
6886  2010.0    99     Furanyl/3-Furanyl fentanyl         0.0   
6887  2010.0    99         Methoxyacetyl fentanyl         0.0   
6888  2010.0    99           Cyclopentyl fentanyl         0.0   
6889  2010.0    99                        U-48800         0.0   
6890  2010.0    99                        U-49900         0.0   
6891  2010.0    99                 Benzylfentanyl         0.0   
6892  2010.0    99              Crotonyl fentanyl         0.0   
6893  2010.0    99                Phenyl fentanyl         0.0   
6894  2010.0    99  Cyclopropyl/Crotonyl Fentanyl         0.0   
6895  2010.0    99                        U-51754         0.0   
6896  2010.0    99            Isobutyryl fentanyl         0.0   
6897  2010.0    99               4-Methylfentanyl         0.0   
6898  2010.0    99     3,4-Methylenedioxy U-47700         0.0   
6899  2010.0    99                     Meta

In [43]:
writer = pd.ExcelWriter('output.xlsx')
data_2010_zone.to_excel(writer, 'Sheet1')
writer.save()

In [61]:
i=0

In [72]:
for i in range(100):
    j=69*i
    data_single_zone=data_2010_zone.iloc[j:(j+69),:]
    #划分为单个zone
    TotalDrugReportZone=sum(data_single_zone['DrugReport'])
    data_2010_zone.iloc[j:(j+69),4]=TotalDrugReportZone

In [73]:
data_2010_zone

YYYY  Zone                  substanceName  DrugReport  \
0     2010.0     0                    Hydrocodone        96.0   
1     2010.0     0                      Methadone        34.0   
2     2010.0     0                      Oxycodone       432.0   
3     2010.0     0                    Oxymorphone         3.0   
4     2010.0     0                  Buprenorphine        47.0   
5     2010.0     0                       Morphine        30.0   
6     2010.0     0                  Hydromorphone        42.0   
7     2010.0     0                     Meperidine         1.0   
8     2010.0     0                       Tramadol        13.0   
9     2010.0     0                        Codeine        17.0   
10    2010.0     0                       Fentanyl         5.0   
11    2010.0     0                         Heroin       304.0   
12    2010.0     0                   Propoxyphene         1.0   
13    2010.0     0                     Methorphan         3.0   
14    2010.0     0                    Pentazocine         0.0   
15    2010.0     0                      Pethidine         0.0   
16    2010.0     0                       Thebaine         0.0   
17    2010.0     0             Dextropropoxyphene         0.0   
18    2010.0     0                        Opiates         0.0   
19    2010.0     0                 Dihydrocodeine         0.0   
20    2010.0     0                Dihydromorphone         0.0   
21    2010.0     0                  Acetylcodeine         0.0   
22    2010.0     0                   Alphaprodine         0.0   
23    2010.0     0                          Opium         1.0   
24    2010.0     0                    Butorphanol         0.0   
25    2010.0     0              Desmethylprodine          0.0   
26    2010.0     0                    Mitragynine         0.0   
27    2010.0     0                           ANPP         0.0   
28    2010.0     0                     Nalbuphine         0.0   
29    2010.0     0                   Remifentanil         0.0   
...      ...   ...                            ...         ...   
6870  2010.0    99                        U-47700         0.0   
6871  2010.0    99                 Fluorofentanyl         0.0   
6872  2010.0    99                 Acryl fentanyl         0.0   
6873  2010.0    99           Cyclopropyl fentanyl         0.0   
6874  2010.0    99               3-Methylfentanyl         0.0   
6875  2010.0    99       Tetrahydrofuran fentanyl         0.0   
6876  2010.0    99               p-Fluorofentanyl         0.0   
6877  2010.0    99    4-Fluoroisobutyryl fentanyl         0.0   
6878  2010.0    99               o-Fluorofentanyl         0.0   
6879  2010.0    99      Fluoroisobutyryl fentanyl         0.0   
6880  2010.0    99               Valeryl fentanyl         0.0   
6881  2010.0    99           cis-3-methylfentanyl         0.0   
6882  2010.0    99      p-methoxybutyryl fentanyl         0.0   
6883  2010.0    99         trans-3-Methylfentanyl         0.0   
6884  2010.0    99               3-Fluorofentanyl         0.0   
6885  2010.0    99        Fluorobutyryl fentanyl          0.0   
6886  2010.0    99     Furanyl/3-Furanyl fentanyl         0.0   
6887  2010.0    99         Methoxyacetyl fentanyl         0.0   
6888  2010.0    99           Cyclopentyl fentanyl         0.0   
6889  2010.0    99                        U-48800         0.0   
6890  2010.0    99                        U-49900         0.0   
6891  2010.0    99                 Benzylfentanyl         0.0   
6892  2010.0    99              Crotonyl fentanyl         0.0   
6893  2010.0    99                Phenyl fentanyl         0.0   
6894  2010.0    99  Cyclopropyl/Crotonyl Fentanyl         0.0   
6895  2010.0    99                        U-51754         0.0   
6896  2010.0    99            Isobutyryl fentanyl         0.0   
6897  2010.0    99               4-Methylfentanyl         0.0   
6898  2010.0    99     3,4-Methylenedioxy U-47700         0.0   
6899  2010.0    99                     Meta

In [74]:
writer = pd.ExcelWriter('output.xlsx')
data_2010_zone.to_excel(writer, 'Sheet1')
writer.save()